### ISSUE TITLE
[FEATURE] Prepare an ADR on type annotations

### ISSUE THREAD

**Issue by konnov** [informalsystems/apalache#162]
The [current approach](https://github.com/informalsystems/apalache/blob/master/docs/types-and-annotations.md) to type annotations in Apalache is quite hacky, for several reasons:

 * we did not want to write own parser for types and introduced a convention, which is okayish but not natural
 * we wanted to wrap problematic expressions such as `{}` and `<<>>` instead of declaring the types of variables, constants, operators, and bound variables.

With the [new type inference](https://github.com/informalsystems/apalache/tree/types) by Jure @Kukovec (to be integrated), we have to find a good way to integrate type information into TLA+ code without touching SANY (the TLA+ parser). Keeping type information on a side, e.g., in a separate file, is not a nice solution.

I see the following approach. We write our own syntax for the type system (actually, we have one already, just no parser yet) and annotate constants, variables, operators and bound names as in the following example:

```tla
EXTENDS Sequences
CONSTANTS Range
Range_type == "set(z)"

VARIABLES list
list_type == "seq(z)"

Mem(e, es) ==
  LET Mem_type == "<a, seq(a)> => Bool" IN
  \E i \in DOMAIN es:
    LET i_type == "Int" IN
    e = es[i]

Init ==
  LET Init_type == "<> => Bool" IN
  list = <<>>

Next ==
  LET Next_type == "<> => Bool" IN
  \E e \in Range:
    LET e_type == "set(z)" IN
    list' = Append(list, e)
```

It is again "convention-over-configuration". Note that types are specified as strings. What I like about it is that operators are easy to write and they are available almost everywhere. However, we should think about type annotations for set comprehensions.

**Comment by lemmy** [comment]
Just a general question: Why is touching SANY a no-go?

**Comment by konnov** [comment]
We are still in the exploration phase and it would be a bad idea to change the syntax all the time. As far as I understand, Leslie also wants to keep the syntax minimalistic (kind of).

**Comment by lemmy** [comment]
(I assume that the ratio of type annotations to the actual behavior spec is constant in the size of the spec)

The examples above don't really look minimalistic from a reader's perspective. Our eyes haven't been trained to filter types in TLA+ specs, but the type annotations make it more difficult to understand the true meaning of the spec.  I assume this is Leslie's main reason to keep the language minimalistic.  As a tool developer, I'm worried that types-as-strings will make it hard for IDEs or end up having poor ergonomics.  By the way, how does this work with `ASSUME` and the adopted best practice of TypeOK invariants?  Will it be the user's responsibility to keep them in sync?

**Comment by konnov** [comment]
The purpose of these type annotations is not to have them written by the user. Though it should be possible. The goal is to have transparent type information (maybe inferred by different tools) right in the spec. Once we have integrated our new type inference, every tool developer will be able to use them, without extending the TLA+ parser or using an external library. The current approach to developing all analyses in a black box makes it harder to integrate tools at all levels. On second thought, the approach presented here is not that different from Pluscal, where the whole code is written in a huge comment.

Think of these types like types in python. If you really want to write the type annotations, you can. But, ideally, it should be done by a type inference engine.

I think there is really a difference between `TypeOK`, which is a simple invariant on variable values, and types that are needed to do any kind of static analysis, translation, etc. Perhaps, these latter types do not add much value to the understanding of a specification, but they are absolutely essential for tooling.

**Comment by ebuchman** [comment]
So tools would be expected to auto update the specs with inferred types? And/or does the user need to provide some hints?  

Why not follow the Pluscal convention and have these types in comments?

And could you give an example of how these type annotations would differ from a TypeOk ?

**Comment by konnov** [comment]
I gave it a bit of thought. Ideally, the authors don't have to write type annotations at all, like they (almost never) do in OCaml or Isabelle. So the tools should infer the types, as they need them. However, there are three reasons why, I believe, the types must be transparent:

1. It is much easier for the user to understand which types are assigned to TLA+ expressions by inspecting an annotated TLA+ file.
2. If somebody is not happy with our type inference algorithm, they can write one of their own and produce a compatible TLA+ file.
3. Our tools are still buggy. For instance, I know exactly how to fix the problem in #157 with a user annotation (in that case, we are talking about assignments, not types). But instead, we have to wait until the bug in the assignment analyzer is fixed. As a user, I am extremely annoyed by the inability to hack away the problem.

If we are talking about an ideal scenario, I would like to have something like [Java annotations](https://docs.oracle.com/javase/1.5.0/docs/guide/language/annotations.html) or [LLVM intrinsics](https://llvm.org/docs/LangRef.html#intrinsic-functions), that is, a way to store metadata right in the code, but without introducing additional code constructs, as metadata depends on the tool. I suspect that it is quite a bit of work on the SANY parser to introduce Java-like annotations. @lemmy, what is your opinion?

As for annotations in the comments, there are two problems:

1. If we are talking about annotations in one large comment, there is no way to refer to an individual operator/variable name. We solve this problem in Apalache by renaming operators and variables, so every name is unique. But obviously this is too much to ask from the user.
2.  I do not see a simple way of writing an annotation preprocessor similar to `javadoc`, as modules can be instantiated. So the preprocessor has to understand the semantics of `INSTANCE`, in order to propagate annotations in the instances.

So the case of Pluscal is simple. You take one language in a large comment and translate it into one TLA+ spec. Here we are talking about type annotations for individual variables, operators, bound variables, etc.

Finally, `TypeOK` is an arbitrary TLA+ predicate. Even we assume that `TypeOK` looks like `A_1 /\ \dots /\ A_n`, where every `A_i` has the form `x_i \in S_i`, there is still plenty of freedom on how one writes `S_i`. For instance, if I write `x \in 1..10`, is `1..10` a type? Some people say "yes", some people say "no". Some people say, it is a refinement type `{i \in Int | 1 <= i <= 10}`. If we don't like to go in undecidable reasoning about types (and I don't) and only want to map expressions to simple types, we better write types as terms. For instance, we have this grammar in the [OOPSLA paper](https://doi.org/10.1145/3360549):

`τ ::= Name | Bool | Int | τ → τ  | Set[τ] | Seq[τ] | τ ∗ ... ∗ τ | [h_1: τ, ..., h_k: τ]`

(It does not work for operators, but let's ignore that fact.) Then, I would like to have an annotation `x_type = "Int"`. Of course, I can write `x \in Int` in `TypeOK`, and that looks fine. But then I like to give a function type from a set of integers to a set of integers, that is, `f_type = "Set[Int] → Set[Int]"`. In `TypeOK`, it is getting ugly: `f \in [SUBSET Int -> SUBSET Int]`. I also suspect that this `TypeOK` also becomes useless for TLC, as it would have to deal with `SUBSET Int`. So in practice, one would write something like `f \in [SUBSET Proc -> SUBSET Proc]`, as the spec most likely talks only about integers from a constant set `Proc`. The [type annotations in Apalache](https://github.com/informalsystems/apalache/blob/master/docs/types-and-annotations.md) are actually so weird because I could not find a nice and concise way of expressing the types that derive from `τ` just by using TLA+ values.

To sum up, `TypeOK` is a useful predicate and it is a good practice to add one. But `TypeOK` inevitably talks about types as sets of values (basically, you have no other choice in TLA+). For the analysis tools, especially the symbolic tools, we care less about the precise sets of values, but we need to know variable sorts (integers, Booleans, sets of integers, sequences of sets of integers).

I guess this is actually the beginning of the ADR. Sorry if the text sounds a bit too aggressive or patronizing. I just had all these things for too long in my head. Now I am happy to have them in this issue :-)

**Comment by konnov** [comment]
```tla
EXTENDS Sequences
CONSTANTS Range
(*                     ^@type("set(z)") *)

VARIABLES list
(*                  ^@type("seq(z)") *)

Mem(e, es) ==
(*
^@type("<a, seq(a)> => Bool") *)
  \E i, j \in DOMAIN es:
(*   ^@type("Int")
         ^@type("Int") *)
    e = es[i]

Init ==
(*
^@type("<> => Bool") *)
  list = <<>>

Next ==
(*@
^@type("<> => Bool") @*)
  \E e \in Range:
(*@^@type("set(z)") @*) 
    list' = Append(list, e)
```

**Comment by konnov** [comment]
That looks completely different from what I expected, but the idea here is that we can introduce annotations graphically right in the comments. Somehow, github preview has a different idea of spaces than the final formatter.

**Comment by konnov** [comment]
Another solution proposed by Leslie Lamport:

```tla
  THEOREM \A a : \A e \in a, es \in Seq(a) : Mem(e, es) \in BOOLEAN
```

> Using the (little-used) TLA+ method of referring to subformulas, this could also be used to provide type declarations for expressions inside definitions.  These type theorems would also be useful in writing proofs.

**Comment by lemmy** [comment]
Assuming I understand you correctly, I think it's a good idea to defer the how of the visualization representation of types to IDEs.  E.g. the Toolbox could show the type annotations on-demand in mouse hovers and basic text editors can show them in-lined.  Pretty-printed specs would simply exclude all type annotations. 

As to writing types when necessary, Leslie's idea of using theorems and the method of referring to subformulas would probably require IDEs to provide refactoring support for (positional) references because - contrary to TLAPS proofs - the spec is unfinished /unstable when a user writes type annotations.  On the other hand, it's probably good advice to use labeled references and rely on positional ones only for "glassbox" specs.

**Comment by konnov** [comment]
I like the approach of using positioned annotations in a comment below an expression. It also gives us a way to introduce all kinds of annotations, not only type annotations. Leslie's idea is interesting, as the theorem could be also proven in TLAPS, if one really wants to do that. But then the types will be far away from the place where they are used.

**Comment by lemmy** [comment]
About types being far away:  The drawback could be alleviated by IDEs showing them in hovers/overlays and commands that navigate the cursor back and forth.

**Comment by konnov** [comment]
As a vim user, I am a bit skeptical about putting too much trust on IDEs :-)

**Comment by lemmy** [comment]
It's 2020 and vim still can do something like overlays/hovers/...? :-p  Teasing aside, TLA+ has several features (PCal mapping, TLAPS annotations, profiling annotations), which only exist in the Toolbox.

Btw. I intentionally wrote IDEs and not "Toolbox" because all of this should be implemented via e.g. [LSP](https://microsoft.github.io/language-server-protocol/) in an IDE agnostic way.

**Comment by konnov** [comment]
> Btw. I intentionally wrote IDEs and not "Toolbox" because all of this should be implemented via e.g. LSP in an IDE agnostic way.

Do you think it makes sense to have TLA+ on the LSP list?

Btw, there are two kinds of annotations: the user annotations (rare but possible) and output annotations. I see how LSP would work for output annotations. Leslie's approach works for both. The approach with comments also works for both, though the formatting is a bit of pain.

**Comment by konnov** [comment]
I guess, vim's approach is to use folding/unfolding, which works well with comments :)

**Comment by konnov** [comment]
LSP looks interesting. Maybe we do not need type annotations tightly integrated into the source code.

@shonfeder, you mentioned LSP in other contexts several times. What is your take on that?

**Comment by lemmy** [comment]
Taking a step back, I wonder if it is really helpful to distinguish between user and type annotations. Won't users just care for what the type inference engine makes out of user-provided annotations?  In other words, why not only ever show output annotations regardless from where the type inference engine sourced its information. Taking another step back, I would probably prefer to read the result of type inference as TLA+ expressions instead of a new syntax.

What LSP list are you referring to (note that there exists no LSP yet)?  LSP is not a proposal on its own, but just a suggestion on how to implement any proposal in a way that is sufficiently IDE agnostic to allow for making IDE features part of a proposal.

**Comment by konnov** [comment]
> What LSP list are you referring to (note that there exists no LSP for TLA+ yet)? 

That is what I meant. TLA+ is not on that [list](https://microsoft.github.io/language-server-protocol/implementors/servers/) ;-)

**Comment by konnov** [comment]
> <a, seq(a)> => Bool

This is an operator type. It takes two arguments (a type variable `a` and a sequence of `a` elements) and returns a Boolean.

**Comment by lemmy** [comment]
It looked at implementing an LSP for TLA+ last summer and there are many use cases for when an LSP would be useful.  Perhaps, types are the tipping point to finally implement an LSP.

**Comment by konnov** [comment]
> Taking another step back, I would probably prefer to read the result of type inference as TLA+ expressions instead of a new syntax

I understand that wish, but there are two problems: (1) typically type grammars are more concise than the types-as-sets syntax, (2) it is not clear how to nicely express some types, e.g., the type of a function from Int to Int, or an operator over an Int and a high-level operator returning a Boolean.

**Comment by lemmy** [comment]
> > <a, seq(a)> => Bool
> 
> This is an operator type. It takes two arguments (a type variable `a` and a sequence of `a` elements) and returns a Boolean.

What I meant is that its new syntax that increases the mental burden, which, perhaps, is fine if I only have to read it rarely but probably an obstacle if I have to type it (especially infrequently).

**Comment by konnov** [comment]
Sure, but if we put lots of constraints on how to write types in TLA+ syntax, it is also... syntax. Using TLA+ does not automatically mean that you have to learn nothing.

**Comment by andrey-kuprianov** [comment]
Here is my proposal wrt. type annotations. I think it would be nice when the type annotation are not simply strings, but TLA+ constructs: that way a huge number of errors will be caught already by the TLA+ syntax checker. Besides, this avoids the problem of a separate syntax inside syntax. The proposed syntax builds on the currently used by Apalache syntax for type annotations with <: operator, which I personally find quite appealing. I have the feeling that with such approach we can even arrive at a solution when type errors could be caught by a model checker / theorem prover -- please correct me if I am wrong. But in any case, these annotations should provide a nice input for the type inference engine -- it will not need to parse them separately from comments, they will just arrive as the integral part of the input file.

```TLA+
---------------------------- MODULE types --------------------------

EXTENDS Sequences, Integers

(* We specify types inside of TLA+, and with the means provided by TLA+.
   For that a user can provide type declarations of the form 
   decl_t, with the meaning is that decl_t returns the type of decl.
   
   The type declarations can be parameterized by other types.
*)


(* This is the syntax for type annotations *)

(* The |= operator returns the type of expression expr; 
   to be used to define types in type declarations
 *)
type |= expr  == type


(* The =| operator specifies the type of expr inline;
   To be used inside other constructs as conjuncts.
*)
expr =| type == TRUE


CONSTANTS Range

VARIABLES list 

Mem(e, es) ==
  \E i, j \in DOMAIN es:
    i =| Int /\ j =| Int /\
    e = es[i]

Init ==
  list = <<>>

Next ==
  \E e \in Range:
    list' = Append(list, e)
    
(* Everything above is just standard TLA+, 
   with the exception of two type annotations for i and j.
   Everything below are type declarations. 
   When needed for understanding, 
   the standard TLA+ of a spec can be intermixed with the type declarations. 
*)

list_t(Z) == 
  Seq(Z) |= list

Mem_t(Z) == 
  \A e \in Z :
  \A es \in list_t(Z) :
  [ Z \X list_t(Z) -> BOOLEAN] |= Mem(e, es)


Init_t == BOOLEAN |= Init

Append_t(Z) ==
  \A e \in Z :
  \A es \in list_t(Z) :
  [ list_t(Z) \X Z -> list_t(Z) ] |= Append(es, e)
  
Next_t == BOOLEAN |= Next

=============================================================================

```

**Comment by shonfeder** [comment]
> @shonfeder, you mentioned LSP in other contexts several times. What is your take on that?

I'm very much pro LSP. I have had a great experience IDE experience (in emacs) with languages that implement the LSP, and LSP seems like its pretty clearly the way forward for editor agnostic (and maybe even application agnostic?) intelligence on analyzing PLs. Also, seems like LSP may neutralize the "but vim isn't an IDE" problem? ;) https://www.kieranbamforth.me/post/vim-lsp

---

Since I was tagged in here, I'll exploit the opening to pitch in my own two cents, from the perspective a currently onboarding developer/user (please counter balance those cents with sufficient grains of salt to account for my newness):

The proposed combination of the `_type` suffix to indicate ascriptions *and* types in strings has the feeling of two orthogonal hacks colliding. On a purely aesthetic level, it's pretty off putting to me.

@andrey-kuprianov's suggestions is the most appealing to me for the inline proposals so far. I'd even propose providing a `Typing` module (akin to the `typing` module supplied by python for its gradual typing) that include those operator definitions and perhaps some convenient types aliases and definitions. (I'm not sold on those particular operators, but think it's much nicer than 

Following @lemmy's good points re: offloading the work to editor integrations, I actually think just using a separate file as the default place for type ascriptions could be quite nice. But I also think interface files can be a nice feature. :)

In my experience, even a simple, non-polymorphic type system can be a very nice aid to thought and development, so I would think its well worth while to invest thought an implementation time into syntax, support, and tooling that will help users feel good about reading and writing the type ascriptions.

**Comment by lemmy** [comment]
> Following @lemmy's good points re: offloading the work to editor integrations, I actually think just using a separate file as the default place for type ascriptions could be quite nice. But I also think interface files can be a nice feature. :)

FWIW: It doesn't have to be a separate file but could simply follow the TLA+ module in the same .tla file (I e.g. recently introduced a [CONFIG construct into TLC](https://github.com/tlaplus/tlaplus/blob/master/tlatools/org.lamport.tlatools/test-model/Alias.tla)):

```tla
---- MODULE Spec ----
...
=====

---- TYPES Spec ----
....
======
```

**Comment by andrey-kuprianov** [comment]
@lemmy , re: 
> FWIW: It doesn't have to be a separate file but could simply follow the TLA+ module in the same .tla file (I e.g. recently introduced a CONFIG construct into TLC)

The solution I proposed was under the assumtion that we can't change the language itself; if we can -- that's a completely different story. E.g., in that case I woud introduce smth like `typeof` construct, i.e. an operator which you could define multiple times for different declarations: `typeof(Mem) == ...`, `typeof(list) == ...`. And a separate construct for type annotations inline, like a colon: `\E i: Int, j: Int \in DOMAIN es...` But, as far as I understand, it's not possible, unfortunately, or am I wrong?

**Comment by konnov** [comment]
This thread is getting more and more detailed. So I have created a PR: https://github.com/informalsystems/apalache/pull/179

Fell free to comment here and there

**Comment by konnov** [comment]
In sum, the `TypeOK` approach does not seem to work for higher-order operators, as operators are not values (surprise!). The term-based approach seems to work, though I am not completely happy about it.

**Comment by lemmy** [comment]
Leslie's comment that "These type theorems would also be useful in writing proofs." paired with his statement that he is more open to amending the proof language, might be another direction that's worth exploring.

**Comment by konnov** [comment]
Right. Actually, types become theorems, once a type inference engine has found them. If we are talking about user annotations, they are more like ASSUME statements, which does not change much, as the syntax may be the same. I will extend the ADR draft. The real stumbling point in the pure TLA+ syntax seems to be the higher-order operators.

**Comment by lemmy** [comment]
> The solution I proposed was under the assumtion that we can't change the language itself; if we can -- that's a completely different story. E.g., in that case I woud introduce smth like `typeof` construct, i.e. an operator which you could define multiple times for different declarations: `typeof(Mem) == ...`, `typeof(list) == ...`. And a separate construct for type annotations inline, like a colon: `\E i: Int, j: Int \in DOMAIN es...` But, as far as I understand, it's not possible, unfortunately, or am I wrong?

@andrey-kuprianov To clarify, I wasn't suggesting to change the language.  I just wanted to hint at the technical fact that SANY poses no restrictions on what precedes or follows the module definition in a `.tla` file.

**Comment by konnov** [comment]
I have just added the section on types-as-strings, so there are three solutions to encoding types (`TypeOK` having a problem with higher-order operators).

**Comment by konnov** [comment]
I have added the section on four approaches to type annotations, extracted from our discussions.  RFC :-)

**Comment by konnov** [comment]
Having 3 options of writing types and 4 options of writing type annotations, this gives us 12 options.

**Comment by konnov** [comment]
I just realized that our type semantics is not always precisely set membership. The reason is, of course, records!

For instance, if we write:

```tla
S == {[type |-> "A", a |-> "Foo"], [type |-> "B", b |-> 123]}
```

Then the expected type of `S` is a set of records that can be extended to the records that belong to the set `[type: STRING, a: STRING, b: STRING]`. So the following statement would be simply incorrect:

```tla
S \in SUBSET [type: STRING, a: STRING, b: STRING]
```

This strange behavior of record types is caused by the need to store records of compatible types in sets. An alternative approach would be to introduce sum types. Then, the type of `S` would be precise, that is:

```tla
S \in (SUBSET [type: STRING, a: STRING]) \union (SUBSET [type: STRING, b: Int])
```

However, this poses another problem. What would be the type of `x` in the following expression?

```tla
\E x \in S:
   ...
```

We do not have a good answer to that, as TLA+ does not have type deconstructors, similar to, e.g., `match` in Scala.

So I am now even more in favor of using a special operator `<:`and types-as-strings instead of set membership and types-as-sets:

```tla
S <: "set([type: Str, a: Str, b: Str])"
```

This notation would tell the reader to understand `S` as a type that can be extended to a set of records `SUBSET [type: STRING, a: STRING, b: STRING]` (by using the subtype relation). That might sound a bit too complicated, but the records in TLA+ are complicated.

**Comment by konnov** [comment]
For the operators, we could also use the inverse notation, as suggested by @shonfeder, e.g.:

```tla
A(x, y) == "(Int, Int) => Int" :>
  x + y
```

This is actually quite similar to Isabelle/HOL. We just use `:>` instead of `where`. (We could even think of writing inductive definitions by using `CASE`.)

**Comment by shonfeder** [comment]
Good catch re: typing records! fwiw, string expressions inside a special operator feel pretty good to me.

**Comment by andrey-kuprianov** [comment]
> ```tla
> A(x, y) == "(Int, Int) => Int" :>
>   x + y
> ```

As a potential user writing type annotations I kind of like that approach -- it looks pretty neat :)

**Comment by konnov** [comment]
I just realized that SANY rejects assumptions over state variables, so we cannot write e.g. `ASSUME(x <: "Str")` for a `VARIABLE x`. This kills yet another neat solution :(

**Comment by konnov** [comment]
It looks like type annotations in comments are the most stable solution after all. That is, points 1.3 and 2.3 in ADR1 #179. Alternatively, we can use my hack, that is, a combination of 1.3 and 2.4, but I agree that it more of a hack.

**Comment by shonfeder** [comment]
Oh well... at least that helps make the decision easier!

**Comment by konnov** [comment]
It's not easy to extract annotations from the comments unless we really integrate with the SANY lexer. So we might still consider the combination of `<:`, `:>` and a type axiom, similar to `TypeOK`:

```tla
TypeAxiom ==
  /\ RM <: "Set(RM)"
  /\ rmState <: "RM -> Str"
  /\ tmState <: "Str"
  /\ tmPrepared <: "Set(RM)"
  /\ msgs <: "Set([type: Str, rm: RM])"
```

In this example, `RM` is a global type variable, which can be instantiated with any type. I think this solution is conceptually much simpler than writing an annotation preprocessor.

**Comment by lemmy** [comment]
`TypeAxiom` looks similar to what an earlier prototype of a symbolic variant of TLC ("BTLC") intended to use: https://github.com/tlaplus/tlaplus/blob/ab14a33e39c78e4c88e81b664b9a8c916b943cab/tlatools/org.lamport.tlatools/src/tlc2/tool/impl/ModelConfig.java#L63-L64

**Comment by konnov** [comment]
I was actually quite surprised that one can only write `ASSUME` over constants. This is in contrast with the assume/assert in model checkers.

**Comment by lemmy** [comment]
TLC!Assert

**Comment by konnov** [comment]
Right. Normally, one would restrict values with `assume` and check them with `assert`, e.g., Boogie [has it](https://stackoverflow.com/questions/42986017/boogie-strange-assertfalse-behavior). TLC has `assert`, but TLA+'s `ASSUME` is quite restricted.

**Comment by konnov** [comment]
I have renamed the previous document to [001rfc-types.md](https://github.com/informalsystems/apalache/blob/4be3c489587c1d9200b93630340acf0a05afa717/docs/adr/001rfc-types.md), as it contains all available options.

The actual ADR [docs/adr/002adr-types.md](https://github.com/informalsystems/apalache/blob/4be3c489587c1d9200b93630340acf0a05afa717/docs/adr/002adr-types.md) contains a concrete proposal on the type grammar and type annotations. See PR #179. 

I already love these type annotations, even without having any type checker for them :-)

**Comment by konnov** [comment]
Leslie Lamport proposes to use the proofs syntax for the higher-order operators as well:

```tla
THEOREM BarType ==
  ASSUME NEW G(_,_),
         \A x \in Int, y \in STRING : G(x,y) \in Int
         PROVE  Bar(G) \in BOOLEAN
```

Similar to that, we can write a theorem about the type of `Foo`:

```tla
THEOREM FooType ==
  \A a \in Int: \A b \in STRING: Foo(a, b) \in Int
```

This looks like a great way of integrating the output of type inference into a TLA+ specification. Theorems like that could be used for the manual proofs with TLAPS.

**Comment by konnov** [comment]
If we like to write types interactively, that is, in the process of spec writing, [ADR-002](https://github.com/informalsystems/apalache/blob/ce155e7815bfb85a1373a389acc9908779c47693/docs/adr/002adr-types.md) gives us a concise way to do that.

**Comment by konnov** [comment]
Here is a solution to type annotations suggested by Leslie Lamport. The method `SemanticNode.getPreComments` in SANY returns the comments that are immediately preceding an AST node. Since SANY keeps the comments, we can extract a type annotation from the comments.

This is an interesting approach and it probably increases usability in comparison to what we have in ADR-002. For instance, the user could write:

```tla
---- MODULE TypedExample ----
VARIABLE
  \* type: Set(Int)
  S

\* type: (Int) => Bool
Mem(e) == \E x \in S: x = e

\* type: () => Bool
Init ==
  S = {}
...
```

The syntax in the comments is quite minimalistic and still easy to parse. Interestingly, SANY is keeping comments as an array, so it should be quite easy to extract the above annotations from the comments.

**Comment by konnov** [comment]
The ADR has not changed much since the discussion. I find it amazing that the type checker is working exactly as we discussed in this issue. There are two-level of annotations: (1) `AssumeType` and `:>` (soon to be pushed), and (2) in-comment annotations (to be implemented). The annotations of the first kind are needed for the internal tooling, whereas the annotations of the second kind are needed for better UX.

**Comment by konnov** [comment]
Just found that `:>` is also defined by the `TLC` module. We need a different operator in `Typing.tla`, before we move to in-comment annotations.

**Comment by konnov** [comment]
As a temporary solution, I will change `:>` to `##` in the type checker.

**Comment by konnov** [comment]
The ADR has not been changing for several months. Closing the issue.
